### PR TITLE
ANN-based mission success prediction model and API

### PR DESCRIPTION
**Pull request by Satyamgupta2365**
## Description

This pull request introduces the **ANN-based Mission Success Prediction Model** and a **Flask API** to serve the model. The model predicts the likelihood of success for military missions based on key mission parameters such as personnel count, equipment status, mission duration, and risk level.

## Why We Need This
- **Better Decision Support**: Data-driven insights for mission planning
- **Resource Optimization**: Optimize personnel and equipment allocation
- **Risk Reduction**: Identify high-risk missions before deployment
- **Utilize Historical Data**: Turn past mission data into predictive intelligence

## Key Features
- **Neural Network Model**: 87% accuracy on test data
- **RESTful API**: Simple JSON interface with authentication
- **Input Parameters**: Unit type, personnel count, equipment status, duration, location risk
- **Output**: Mission success probability

### Flask API
- A simple **RESTful API** designed to serve the trained model
- Accepts mission data in JSON format
- Processes input through the model
- Returns mission success probability

### Requirements File
- `requirements.txt` file to easily set up Python dependencies
- Includes essential libraries like Flask, NumPy, TensorFlow, etc.

### Sample Data
- `mission_data_sample.xlsx`: Contains 100 sample records for mission data
- Used for training and testing the predictive model

### 1. Clone the Repository
```bash
git clone https://github.com/Satyamgupta2365/War-Data-Analysis
cd War-Data-Analysis/mission_predictor
```

### 2. Install Dependencies
Ensure Python is installed and run:
```bash
pip install -r requirements.txt
```

### 3. Start the Flask API
Run the Flask app to start serving the model:
```bash
python mission_predictor_api.py
```

### 4. Send Prediction Requests
Use curl to send POST requests to the Flask API:
```bash
curl -X POST http://localhost:5000/predict \
     -H "Content-Type: application/json" \
     -d '{"unit_type": 1, "personnel_count": 250, "equipment_operational_ratio": 0.8, "duration_days": 14, "location_risk": 2}'
```

### Example Response
```json
{
    "success_probability": 0.87923
}
```

## Next Steps
- **Integration**: Integrate this prediction model into the main War Management System for live data predictions
- **Model Expansion**: 
  - Explore adding more mission-related features
  - Improve prediction accuracy
  - Train the model on more diverse data

## Technical Details
- **Model Type**: Artificial Neural Network (ANN)
- **Training Framework**: TensorFlow
- **Preprocessing**: Scikit-learn
- **API Framework**: Flask
- **Data Source**: Historical mission records

## Contributions
- Developed predictive model for mission success
- Created scalable Flask API for model inference
- Provided comprehensive documentation and usage instructions

## Potential Impact
This model can assist military strategists in:
- Risk assessment
- Mission planning
- Resource allocation
- Predictive decision-making

### CODE REVIEW

**Review by itsmenuma — CHANGES_REQUESTED** [review]
Thank you for your contribution! This is a well-structured and thoughtful addition to the repository. The ANN-based mission success prediction model and accompanying Flask API provide significant value, especially with the included sample data and clear instructions for usage.

Initial Feedback:

✅ Great points:

Well-documented usage instructions

Proper separation of concerns (model training vs. API serving)

Inclusion of sample dataset and requirements.txt

🔍 Suggestions for improvement before merging:

Security Consideration: Ensure user input is validated/sanitized in the API to prevent misuse.

Model Loading: Consider saving and loading the trained model using model.save() / load_model() instead of re-training on every run (if that's currently the case).

API Error Handling: Add error handling for malformed JSON or missing fields in the request.

Testing: Add a basic unit test or integration test to validate the API response under expected input conditions.

Data Privacy: Make sure that the mission_data_sample.xlsx doesn’t contain any sensitive or realistic military data. If it's mock data, consider renaming to clarify (e.g., mock_mission_data.xlsx).

Once these minor improvements are addressed, I’ll be happy to proceed with merging. Great job!

**Comment by Satyamgupta2365** [comment]
Hi ,
First a fall thanks for feedback
I wanted to let you know that I've successfully implemented the
improvements suggested for the Mission Success Predictor module in the War
Management System repository. The updates include:

✅ Secure input validation for the Flask API
✅ Error handling for missing or malformed JSON requests
✅ Scaler persistence using numpy arrays
✅ Saved model loading to avoid retraining on every startup
✅ Clear naming for the dataset file (`mock_mission_data.xlsx`) to indicate
that it contains mock data
✅ A clean and modular structure for easier integration and testing

All of these changes are now reflected in the updated pull request. Please
feel free to review and suggest any further refinements.
Looking forward to your feedback and hoping this can be merged soon!

Best regards,
Satyam Gupta

**Review by itsmenuma — APPROVED** [review]
Hi Satyam,

Thanks again for implementing all the feedback so effectively — the improvements really enhance the robustness and usability of the Mission Success Predictor module. 🙌 I'm approving and merging the PR. Great work!

If you're open to exploring further improvements, here are a few optional suggestions for the future:

🔄 Possible Future Enhancements:
Unit Tests for the Flask API
Adding basic unit or integration tests (e.g., using pytest + Flask's test client) will help ensure endpoints behave as expected, especially after future changes.

Input Schema Validation
Using libraries like pydantic or marshmallow can help enforce data types and provide clearer validation errors.

Dockerization (Optional)
Packaging the app into a Docker container could make deployment easier and more portable across environments.

Logging Improvements
Adding logging (using Python’s logging module) instead of relying on print statements would make the app more production-ready and help with debugging in live environments.

Model Versioning/Tracking
If you plan on training and deploying updated models in the future, consider using a simple versioning system or tools like MLflow for better model lifecycle management.

These are just ideas to explore over time — no urgency at all. Your current implementation already looks great and is a strong foundation! 💪

And once again, if you're enjoying the project, a ⭐ on the repo is always appreciated. 😊

Looking forward to more awesome contributions from you!

Best regards,
Numa
